enterprise_access/apps/content_assignments/tests/test_api.py

```diff
@@ -366,3 +374,151 @@ def test_cancel_assignments_happy_path(self):
+            'title': content_title,
+        }
+
+        if user_exists:
+            UserFactory(username='alice', email=learner_email, lms_user_id=lms_user_id)
+
+        assignment = None
+        if existing_assignment_state:
+            assignment = LearnerContentAssignmentFactory.create(
+                assignment_configuration=self.assignment_configuration,
+                learner_email=learner_email,
+                lms_user_id=None,
+                content_key=content_key,
+                content_title=content_title,
+                content_quantity=-content_price_cents,
+                state=existing_assignment_state,
+            )
+
+        allocate_assignments(
+            self.assignment_configuration,
+            [learner_email],
+            content_key,
+            content_price_cents,
+        )
+
+        # Get the latest assignment from the db.
+        assignment = LearnerContentAssignment.objects.get(learner_email=learner_email)
+
+        # We should have updated the lms_user_id of the assignment IFF a user pre-existed.
+        if user_exists:
+            assert assignment.lms_user_id == lms_user_id
+        else:
+            assert assignment.lms_user_id is None
+
+    @mock.patch(
+        'enterprise_access.apps.content_assignments.api.USER_EMAIL_READ_BATCH_SIZE',
+        TEST_USER_EMAIL_READ_BATCH_SIZE,
+    )
+    def test_try_populate_assignments_lms_user_id_with_batching(self):
+        """
+        Tests the _try_populate_assignments_lms_user_id() function, especially the read batching feature.
+
+        Make 2*N users with unique emails,
+        Call function on:
+            2*N assignments with existing emails,
+            and 0.5*N assignment with new emails,
+        Where N is the max batch size.
+        """
+        num_users_to_create = 2 * TEST_USER_EMAIL_READ_BATCH_SIZE
+        existing_users = [UserFactory() for _ in range(num_users_to_create)]
+
+        assignments_for_non_existing_users = [
+            LearnerContentAssignmentFactory.create(
+                assignment_configuration=self.assignment_configuration,
+                lms_user_id=None,
+                state=LearnerContentAssignmentStateChoices.ALLOCATED,
+            )
+            for _ in range(int(0.5 * TEST_USER_EMAIL_READ_BATCH_SIZE))
+        ]
+        assignments_for_existing_users = [
+            LearnerContentAssignmentFactory.create(
+                assignment_configuration=self.assignment_configuration,
+                learner_email=existing_user.email,
+                lms_user_id=None,
+                state=LearnerContentAssignmentStateChoices.ALLOCATED,
+            )
+            for existing_user in existing_users
+        ]
+        # Assemble a list of assignments 2.5 longer than the max batch size.
+        all_assignments = assignments_for_non_existing_users + assignments_for_existing_users
+
+        _try_populate_assignments_lms_user_id(all_assignments)
```

Comment:
minor suggestion: this is one of those rare cases where `assertNumQueries` might actually be valuable. https://docs.djangoproject.com/en/4.2/topics/testing/tools/#django.test.TransactionTestCase.assertNumQueries